### PR TITLE
feat: Define new History format and message organization

### PR DESCRIPTION
**Pull request by kgilpin**
Related to: 

* https://github.com/orgs/getappmap/projects/15?pane=issue&itemId=74967761
* https://github.com/getappmap/vscode-appland/pull/1006

## Analysis

1. **Unify the history management** into a centralized and structured framework.
2. **Introduce new history functionalities** to better retrieve and store interaction data.
3. **Add migration support** to seamlessly transition from the old history format to the new format.
4. **Enhance RPC support** for better integration and communication with other system components.

## Proposed Changes

1. **Command Handler (rpc.ts)**:
   - Import and integrate new history helpers and handlers.
   - Include initialization and migration logic for the history system in the command handler.

2. **Explain RPC Handlers (explain.ts)**:
   - Add a new `loadThread` function to handle fetching of specific threads.
   - Export new handler functions for explain commands.

3. **History Management (history.ts)**:
   - Define a new `History` class encapsulating methods to manage questions, tokens, and answers.
   - Implement logic for loading history, organizing data, and handling errors.
   - Add a `migrate` method to transition from old history format to the new format.

4. **Local Navie Integration (navie-local.ts)**:
   - Replace the old local history management logic with the new centralized history mechanism.
   - Utilize the new `History` class along with its methods to manage conversation threads.

5. **Remote Navie Integration (navie-remote.ts)**:
   - Introduce a callback handler for managing various events such as acknowledgment, token reception, context requests, and completion using the new history system.
   - Modify the ask method to use the new remote callback handler for history operations.

6. **Components Integration (AppMapRPC.ts and ChatSearch.vue)**:
    - Update components to interface with the new RPC methods for loading threads and managing history state in the UI.

## History storage design

```
history/
|-- threads/
|   |-- <threadId>/
|       |-- messages/
|       |   |-- <userMessageId>/
|       |   |   |-- question.txt
|       |   |   |-- codeSelection.txt
|       |   |   |-- prompt.md
|       |   |   |-- assistantMessageId.txt
|       |   |   |-- answer.md
|       |-- sequence/
|       |   |-- <timestamp>.<field>.<extension> (symlinks to ../messages/<userMessageId>/<field>.<extension>)
|       |-- projectDirectories.txt
|-- dates/
|   |-- <YYYY-MM-DD>/
|       |-- <threadId> (symlink to ../threads/<threadId>/)
```

### Explanation:

- **`history/`**: The root directory containing all history-related data.
- **`threads/`**: Subdirectory containing individual thread directories.
  - **`<threadId>/`**: Directory for each thread identified by its unique thread ID.
    - **`messages/`**: Subdirectory containing user and assistant messages.
      - **`<userMessageId>/`**: Subdirectory containing message files for a specific user message.
        - **`question.txt`**: File containing the user's question.
        - **`codeSelection.txt`**: Optional file containing code selection information.
        - **`prompt.md`**: Optional file containing any prompts.
        - **`assistantMessageId.txt`**: File containing the assistant message ID.
        - **`answer.md`**: File containing the assistant's answer.
    - **`sequence/`**: Subdirectory containing sequence files.
      - **`<timestamp>.<field>.<extension>`**: Symlinks to the corresponding message files in the messages directory.
    - **`projectDirectories.txt`**: File containing the list of project directories associated with the thread.
- **`dates/`**: Subdirectory categorizing threads by date.
  - **`<YYYY-MM-DD>/`**: Directory for each date.
    - **`<threadId>`**: Symlink to the thread directory under `threads/`.

This structure allows for easy access and sorting of messages and threads based on their timestamps and other criteria.

## UI Example

Screen Shot 2024-08-20 at 9 06 11 AM

### CODE REVIEW

File: packages/components/src/pages/ChatSearch.vue [hunk 1]
```diff
@@ -254,6 +250,24 @@ export default {
...
+      const thread = await this.rpcClient.loadThread(threadId);
+      const chat = this.$refs.vchat;
+
+      for (const message of thread.messages) {
+        if (message.role === 'user') {
+          const userMessage = chat.addUserMessage(message);
+          if (message.codeSelection) userMessage.codeSelections = [message.codeSelection];
+          // TODO: User message provides prompt, but the UI does not have a place for this.
+        } else {
+          const systemMessage = chat.addSystemMessage(message);
```
Comment (dustinbyrne) [inline packages/components/src/pages/ChatSearch.vue:265]: Double check the arguments here for `addUserMessage`/`addSystemMessage`. I don't think they accept a full message structure, though I wouldn't be opposed to that. This is the downside of `//@ts-nocheck`.

File: packages/components/src/pages/ChatSearch.vue [hunk 2]
```diff
@@ -250,6 +250,30 @@ export default {
     },
   },
   methods: {
+    async loadThread(threadId: string) {
+      const thread = await this.rpcClient.loadThread(threadId);
+      const chat = this.$refs.vchat;
```
Comment (dustinbyrne) [inline packages/components/src/pages/ChatSearch.vue:255]: Should `chat.threadId` be set here as well?

Comment (kgilpin) [reply packages/components/src/pages/ChatSearch.vue:255]: Yeah it seems like it should

Comment (kgilpin) [reply packages/components/src/pages/ChatSearch.vue:255]: This is addressed now.

File: packages/cli/src/rpc/explain/navie/history.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        }
+
+        // Project directories are unknown, so overwrite the file contents with empty text.
+        const projectDirectoriesFile = join(
+          history.directory,
+          'threads',
+          threadId,
+          'projectDirectories.txt'
+        );
+        await writeFile(projectDirectoriesFile, '');
```
Comment (dustinbyrne) [inline packages/cli/src/rpc/explain/navie/history.ts:344]: I'm seeing this locally.

```
3399894 [Stderr] [history] Migrating thread 1866418e-0b40-4393-95f2-003b45ebf23e from old history format
3399894 [Stderr] Error: ENOENT: no such file or directory, open '/home/db/.appmap/navie/history/threads/1866418e-0b40-4393-95f2-003b45ebf23e/projectDirectories.txt'
3399894 [Stderr]     at async open (node:internal/fs/promises:639:25)
3399894 [Stderr]     at async writeFile (node:internal/fs/promises:1219:14)
3399894 [Stderr]     at async History.migrate (/home/db/dev/applandinc/appmap-js/packages/cli/built/rpc/explain/navie/history.js:271:17)
3399894 [Stderr]     at async Object.handler (/home/db/dev/applandinc/appmap-js/packages/cli/built/cmds/index/rpc.js:76:9) {
3399894 [Stderr]   errno: -2,
3399894 [Stderr]   code: 'ENOENT',
3399894 [Stderr]   syscall: 'open',
3399894 [Stderr]   path: '/home/db/.appmap/navie/history/threads/1866418e-0b40-4393-95f2-003b45ebf23e/projectDirectories.txt'
3399894 [Stderr] }
3399894 [Stderr] exited with code 1
3399894 [Stderr] too many crashes - aborting
```

Comment (kgilpin) [reply packages/cli/src/rpc/explain/navie/history.ts]: I've added an update for this, it should ignore old history that contains no files or unrecognized roles.

Check thread `1866418e-0b40-4393-95f2-003b45ebf23e` to see if it's either:

a) Empty
b) Does not contain any messages that have either the `user` or `assistant` role.

File: packages/cli/src/rpc/explain/navie/history.ts [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+      threadFiles.sort((a, b) => threadFileAsTimestamp(a) - threadFileAsTimestamp(b));
+
+      let lastUserMessageId: string | undefined;
+      for (const threadFile of threadFiles) {
+        const timestamp = threadFileAsTimestamp(threadFile);
+        // MessageIds are not available, so fake them with timestamps.
+        const messageId = timestamp.toString();
+        const messageFile = join(oldThreadDir, threadFile);
+        const messageStr = await readFile(messageFile, 'utf-8');
+        const message = JSON.parse(messageStr) as Message;
```
Comment (dustinbyrne) [inline packages/cli/src/rpc/explain/navie/history.ts:327]: One final issue here.

I'm seeing cases where this raises an exception. I believe some of these files could have been empty, which would result in `SyntaxError: Unexpected end of JSON input`.

I've lost the files now so I can't recreate this 100%.

**Review by dustinbyrne — APPROVED** [review]
Looking good!

**Pull request merged**